python/cpp/client.cpp

```diff
@@ -0,0 +1,139 @@
+        return socket;
+    }
+}
+}
+
+std::vector<NodeType> query(std::vector<NodeType> in,
+                            std::string app_mode,
+                            int request_timeout = 1500,
+                            int request_retries = 2)
+{
+    zmq::context_t context (1);
+    auto retries_left = int { request_retries };
+
+    auto reply_list = std::vector<NodeType>();
+    auto socket = local::get_socket(context, app_mode);
+
+    auto buffer = std::ostringstream();
+    {
+        boost::archive::text_oarchive archive(buffer);
+        archive << in;
+    }
+    zmq::message_t message((void*)buffer.str().c_str(), buffer.str().size()+1);
+
+    socket->send(message);
+
+    auto expect_reply = bool { true };
+    while (expect_reply){
+        zmq::pollitem_t items[] = {
+                {static_cast<void*>(*socket), 0 , ZMQ_POLLIN, 0}
+        };
+        zmq::poll(&items[0], 1, request_timeout);
+
+        // If there is a reply, process it
+        if (items[0].revents & ZMQ_POLLIN) {
+            zmq::message_t reply;
+            socket->recv(&reply);
+            std::istringstream in_buffer(static_cast<char *>(reply.data()));
+            boost::archive::text_iarchive archive(in_buffer);
+            archive >> reply_list;
+            socket->close();
+            if (reply_list.size() > 0 &&
+                reply_list.at(0).count("error") &&
+                reply_list.at(0).count("action") &&
+                boost::get<std::string>(reply_list.at(0).at("action")) == std::string("resend")) {
+                if (--retries_left == 0) {
+                    reply_list.push_back({{{std::string("error"), std::string("daemon not responding")}}});
+                    expect_reply = false;
+                    continue;
+                }
+                std::cout << "Error: daemon could not deserialize message, retrying" << std::endl;
+                std::cout << "The sent data was " << message.str() << std::endl;
+                reply_list.clear();
+                socket = local::get_socket(context, app_mode);
+                socket->send(message);
+            }
+            else {
+                expect_reply = false;
+            }
+        }
+        else if (--retries_left == 0) {
+            std::cout << "Error: hierarchy daemon does not reply" << std::endl;
+            expect_reply = false;
+            socket->close();
+            socket.reset();
+            reply_list.push_back({{{std::string("error"), std::string("daemon not responding")}}});
+        }
+        else {
+            std::cout << "Warning: no response from daemon, retrying" << std::endl;
+            socket->close();
+            socket.reset();
+            socket = local::get_socket(context, app_mode);
+            socket->send(message);
+        }
+
+    }
+    context.close();
+    return reply_list;
+}
+
+namespace py = pybind11;
+
+PYBIND11_MODULE(tag_hierarchy_client, m)
```

Comment:
Is this the actual binding?

Reply:
Yep, it is where the C++ function gets exposed as a python function. Fix for documentation is incoming.